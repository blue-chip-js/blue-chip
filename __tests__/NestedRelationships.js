import checklistResources from "../__testHelpers__/fixtures/checklistsAndTasksNormalized";
import pOResources from "../__testHelpers__/fixtures/purchaseOrdersAndOthersNormalized";
import specResources from "../__testHelpers__/fixtures/specsNormalized";
import specDetailComsResources from "../__testHelpers__/fixtures/specDetailComsNormalized";
import {
  Checklist,
  PurchaseOrderContact,
  Spec,
  SpecDetailCom,
} from "../__testHelpers__/models";

describe("Nested resources", () => {
  test("includes with hasMany.belongsTo", () => {
    const checklists = Checklist.query(checklistResources)
      .all()
      .includes(["tasks.user"])
      .toObjects();
    expect(checklists).toMatchSnapshot();
  });

  test("includes with belongsTo.belongsTo", () => {
    const purchaseOrderContacts = PurchaseOrderContact.query(pOResources)
      .includes(["vendorContact.contact"])
      .toObjects();
    expect(purchaseOrderContacts).toMatchSnapshot();
  });

  test("belongsTo.belongsTo.belongsTo.belongsTo.belongsTo", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [97]})
      .includes(["specDetail.spec.area.areaType.someNestedResource"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo.belongsTo.belongsTo.belongsTo.belongsTo", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [97]})
      .includes([
        "specDetail.spec.area.areaType.someNestedResource.anotherNestedResource",
      ])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("includes with hasMany.hasMany", () => {
    const specs = Spec.query(specResources)
      .where({id: [11]})
      .includes(["specDetails.coms"])
      .toObjects();
    expect(specs).toMatchSnapshot();
  });

  test("belongsTo, hasMany.[hasMany, hasMany, belongsTo]", () => {
    const specs = Spec.query(specResources)
      .where({id: [11]})
      .includes(["specCategory", "specDetails.[coms, roomTypes, user]"])
      .toObjects();
    expect(specs).toMatchSnapshot();
  });

  test("hasMany.[belongsTo, hasMany, hasMany], belongsTo, switched order", () => {
    const specs = Spec.query(specResources)
      .where({id: [11]})
      .includes(["specDetails.[user, coms, roomTypes]", "specCategory"])
      .toObjects();
    expect(specs).toMatchSnapshot();
  });

  test("belongsTo, hasMany.[hasMany, belongsTo, hasMany], mixed order", () => {
    const specs = Spec.query(specResources)
      .where({id: [11]})
      .includes(["specCategory", "specDetails.[coms, user, roomTypes]"])
      .toObjects();
    expect(specs).toMatchSnapshot();
  });

  test("belongsTo.[belongsTo, belongsTo]", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [99]})
      .includes(["spec.[area, specCategory]"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo, belongsTo.belongsTo", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [99]})
      .includes(["spec.area", "spec.specCategory"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo, belongsTo.hasMany, belongsTo.belongsTo", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [99]})
      .includes(["spec.area", "spec.specDetails", "spec.specCategory"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.[belongsTo, hasMany]", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [99]})
      .includes(["spec.[area, specDetails]"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo.[belongsTo, belongsTo]", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [97]})
      .includes(["specDetail.spec.[area, specCategory]"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo.belongsTo.belongsTo", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [97]})
      .includes(["specDetail.spec.area.areaType"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo.belongsTo.hasMany", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [97]})
      .includes(["specDetail.spec.area.areaRooms"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo.[belongsTo.belongsTo, belongsTo]", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [97]})
      .includes(["specDetail.spec.[area.areaType, specCategory]"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.belongsTo.[belongsTo.hasMany, belongsTo]", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [97]})
      .includes(["specDetail.spec.[area.areaRooms, specCategory]"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.[hasMany, belongsTo], switched order", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [99]})
      .includes(["spec.[specDetails, area]"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("belongsTo.[belongsTo, hasMany, belongsTo], mixed", () => {
    const specDetailComs = SpecDetailCom.query(specDetailComsResources)
      .where({id: [99]})
      .includes(["spec.[specCategory, specDetails, area]"])
      .toObjects();
    expect(specDetailComs).toMatchSnapshot();
  });

  test("hasMany.hasMany.hasMany", () => {
    const specs = Spec.query(specResources)
      .where({id: [11]})
      .includes(["specDetails.coms.tasks"])
      .toObjects();
    expect(specs).toMatchSnapshot();
  });

  test("hasMany.hasMany.hasMany.belongsTo", () => {
    const specs = Spec.query(specResources)
      .where({id: [11]})
      .includes(["specDetails.coms.tasks.user"])
      .toObjects();
    expect(specs).toMatchSnapshot();
  });

  test("hasMany.[hasMany.hasMany.hasMany, hasMany, belongsTo]", () => {
    const specs = Spec.query(specResources)
      .where({id: [11]})
      .includes(["specDetails.[coms.tasks.user, roomTypes, user]"])
      .toObjects();
    expect(specs).toMatchSnapshot();
  });
});
