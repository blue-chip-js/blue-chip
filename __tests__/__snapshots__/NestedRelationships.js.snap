// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nested resources belongsTo, hasMany.[hasMany, belongsTo, hasMany], mixed order 1`] = `
Array [
  Object {
    "id": 11,
    "specCategory": Object {
      "id": 7,
      "specCategoryData": "Should show up",
    },
    "specData": "Should show up",
    "specDetails": Array [
      Object {
        "coms": Array [
          Object {
            "comData": "Should show up",
            "id": 2,
          },
        ],
        "id": 12,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 1",
        "user": Object {
          "id": 5,
          "usersData": "Should show up",
        },
      },
      Object {
        "id": 13,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 2, no coms",
      },
    ],
  },
]
`;

exports[`Nested resources belongsTo, hasMany.[hasMany, hasMany, belongsTo] 1`] = `
Array [
  Object {
    "id": 11,
    "specCategory": Object {
      "id": 7,
      "specCategoryData": "Should show up",
    },
    "specData": "Should show up",
    "specDetails": Array [
      Object {
        "coms": Array [
          Object {
            "comData": "Should show up",
            "id": 2,
          },
        ],
        "id": 12,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 1",
        "user": Object {
          "id": 5,
          "usersData": "Should show up",
        },
      },
      Object {
        "id": 13,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 2, no coms",
      },
    ],
  },
]
`;

exports[`Nested resources belongsTo.[belongsTo, belongsTo] 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 99,
    "spec": Object {
      "area": Object {
        "areasData": "Should show up",
        "id": 51,
      },
      "id": 11,
      "specCategory": Object {
        "id": 7,
        "specCategoryData": "Should show up",
      },
      "specData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.[belongsTo, hasMany, belongsTo], mixed 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 99,
    "spec": Object {
      "area": Object {
        "areasData": "Should show up",
        "id": 51,
      },
      "id": 11,
      "specCategory": Object {
        "id": 7,
        "specCategoryData": "Should show up",
      },
      "specData": "Should show up",
      "specDetails": Array [
        Object {
          "id": 1,
          "specDetailsData": "Should show up 1",
        },
        Object {
          "id": 2,
          "specDetailsData": "Should show up 2",
        },
      ],
    },
  },
]
`;

exports[`Nested resources belongsTo.[belongsTo, hasMany] 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 99,
    "spec": Object {
      "area": Object {
        "areasData": "Should show up",
        "id": 51,
      },
      "id": 11,
      "specData": "Should show up",
      "specDetails": Array [
        Object {
          "id": 1,
          "specDetailsData": "Should show up 1",
        },
        Object {
          "id": 2,
          "specDetailsData": "Should show up 2",
        },
      ],
    },
  },
]
`;

exports[`Nested resources belongsTo.[hasMany, belongsTo], switched order 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 99,
    "spec": Object {
      "area": Object {
        "areasData": "Should show up",
        "id": 51,
      },
      "id": 11,
      "specData": "Should show up",
      "specDetails": Array [
        Object {
          "id": 1,
          "specDetailsData": "Should show up 1",
        },
        Object {
          "id": 2,
          "specDetailsData": "Should show up 2",
        },
      ],
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo, belongsTo.belongsTo 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 99,
    "spec": Object {
      "area": Object {
        "areasData": "Should show up",
        "id": 51,
      },
      "id": 11,
      "specCategory": Object {
        "id": 7,
        "specCategoryData": "Should show up",
      },
      "specData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo, belongsTo.hasMany, belongsTo.belongsTo 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 99,
    "spec": Object {
      "area": Object {
        "areasData": "Should show up",
        "id": 51,
      },
      "id": 11,
      "specCategory": Object {
        "id": 7,
        "specCategoryData": "Should show up",
      },
      "specData": "Should show up",
      "specDetails": Array [
        Object {
          "id": 1,
          "specDetailsData": "Should show up 1",
        },
        Object {
          "id": 2,
          "specDetailsData": "Should show up 2",
        },
      ],
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo.[belongsTo, belongsTo] 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 97,
    "specDetail": Object {
      "id": 5,
      "spec": Object {
        "area": Object {
          "areasData": "Should show up",
          "id": 51,
        },
        "id": 12,
        "specCategory": Object {
          "id": 7,
          "specCategoryData": "Should show up",
        },
        "specData": "Should show up",
      },
      "specDetailsData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo.[belongsTo.belongsTo, belongsTo] 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 97,
    "specDetail": Object {
      "id": 5,
      "spec": Object {
        "area": Object {
          "areaType": Object {
            "areaTypeData": "Should show up",
            "id": 2,
          },
          "areasData": "Should show up",
          "id": 51,
        },
        "id": 12,
        "specCategory": Object {
          "id": 7,
          "specCategoryData": "Should show up",
        },
        "specData": "Should show up",
      },
      "specDetailsData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo.[belongsTo.hasMany, belongsTo] 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 97,
    "specDetail": Object {
      "id": 5,
      "spec": Object {
        "area": Object {
          "areaRooms": Array [
            Object {
              "areaRoomData": "Should show up",
              "id": 13,
            },
            Object {
              "areaRoomData": "Should show up",
              "id": 14,
            },
          ],
          "areasData": "Should show up",
          "id": 51,
        },
        "id": 12,
        "specCategory": Object {
          "id": 7,
          "specCategoryData": "Should show up",
        },
        "specData": "Should show up",
      },
      "specDetailsData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo.belongsTo.belongsTo 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 97,
    "specDetail": Object {
      "id": 5,
      "spec": Object {
        "area": Object {
          "areaType": Object {
            "areaTypeData": "Should show up",
            "id": 2,
          },
          "areasData": "Should show up",
          "id": 51,
        },
        "id": 12,
        "specData": "Should show up",
      },
      "specDetailsData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo.belongsTo.belongsTo.belongsTo 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 97,
    "specDetail": Object {
      "id": 5,
      "spec": Object {
        "area": Object {
          "areaType": Object {
            "areaTypeData": "Should show up",
            "id": 2,
            "someNestedResource": Object {
              "a": 1
            }
          },
          "areasData": "Should show up",
          "id": 51,
        },
        "id": 12,
        "specData": "Should show up",
      },
      "specDetailsData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo.belongsTo.belongsTo.belongsTo.belongsTo 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 97,
    "specDetail": Object {
      "id": 5,
      "spec": Object {
        "area": Object {
          "areaType": Object {
            "areaTypeData": "Should show up",
            "id": 2,
            "someNestedResource": Object {
              "a": 1,
              "anotherNestedResource": Object {
                "b": 2
              }
            }
          },
          "areasData": "Should show up",
          "id": 51,
        },
        "id": 12,
        "specData": "Should show up",
      },
      "specDetailsData": "Should show up",
    },
  },
]
`;

exports[`Nested resources belongsTo.belongsTo.belongsTo.hasMany 1`] = `
Array [
  Object {
    "comData": "Should show up",
    "id": 97,
    "specDetail": Object {
      "id": 5,
      "spec": Object {
        "area": Object {
          "areaRooms": Array [
            Object {
              "areaRoomData": "Should show up",
              "id": 13,
            },
            Object {
              "areaRoomData": "Should show up",
              "id": 14,
            },
          ],
          "areasData": "Should show up",
          "id": 51,
        },
        "id": 12,
        "specData": "Should show up",
      },
      "specDetailsData": "Should show up",
    },
  },
]
`;

exports[`Nested resources hasMany.[belongsTo, hasMany, hasMany], belongsTo, switched order 1`] = `
Array [
  Object {
    "id": 11,
    "specCategory": Object {
      "id": 7,
      "specCategoryData": "Should show up",
    },
    "specData": "Should show up",
    "specDetails": Array [
      Object {
        "coms": Array [
          Object {
            "comData": "Should show up",
            "id": 2,
          },
        ],
        "id": 12,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 1",
        "user": Object {
          "id": 5,
          "usersData": "Should show up",
        },
      },
      Object {
        "id": 13,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 2, no coms",
      },
    ],
  },
]
`;

exports[`Nested resources hasMany.[hasMany.hasMany.hasMany, hasMany, belongsTo] 1`] = `
Array [
  Object {
    "id": 11,
    "specData": "Should show up",
    "specDetails": Array [
      Object {
        "coms": Array [
          Object {
            "comData": "Should show up",
            "id": 2,
            "tasks": Array [
              Object {
                "id": 3,
                "tasksData": "Should show up",
                "user": Object {
                  "id": 5,
                  "usersData": "Should show up",
                },
              },
            ],
          },
        ],
        "id": 12,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 1",
        "user": Object {
          "id": 5,
          "usersData": "Should show up",
        },
      },
      Object {
        "id": 13,
        "roomTypes": Array [
          Object {
            "id": 2,
            "roomTypeData": "Should show up",
          },
          Object {
            "id": 4,
            "roomTypeData": "Should show up",
          },
        ],
        "specDetailData": "Should show up 2, no coms",
      },
    ],
  },
]
`;

exports[`Nested resources hasMany.hasMany.hasMany 1`] = `
Array [
  Object {
    "id": 11,
    "specData": "Should show up",
    "specDetails": Array [
      Object {
        "coms": Array [
          Object {
            "comData": "Should show up",
            "id": 2,
            "tasks": Array [
              Object {
                "id": 3,
                "tasksData": "Should show up",
              },
            ],
          },
        ],
        "id": 12,
        "specDetailData": "Should show up 1",
      },
      Object {
        "id": 13,
        "specDetailData": "Should show up 2, no coms",
      },
    ],
  },
]
`;

exports[`Nested resources hasMany.hasMany.hasMany.belongsTo 1`] = `
Array [
  Object {
    "id": 11,
    "specData": "Should show up",
    "specDetails": Array [
      Object {
        "coms": Array [
          Object {
            "comData": "Should show up",
            "id": 2,
            "tasks": Array [
              Object {
                "id": 3,
                "tasksData": "Should show up",
                "user": Object {
                  "id": 5,
                  "usersData": "Should show up",
                },
              },
            ],
          },
        ],
        "id": 12,
        "specDetailData": "Should show up 1",
      },
      Object {
        "id": 13,
        "specDetailData": "Should show up 2, no coms",
      },
    ],
  },
]
`;

exports[`Nested resources includes with belongsTo.belongsTo 1`] = `
Array [
  Object {
    "id": 13,
    "pocData": "f",
    "vendorContact": Object {
      "contact": Object {
        "contactData": "a",
        "id": 12,
      },
      "id": 15,
      "vcData": "b",
    },
  },
]
`;

exports[`Nested resources includes with hasMany.belongsTo 1`] = `
Array [
  Object {
    "id": 3,
    "name": "QA Checklist Rest",
    "tasks": Array [
      Object {
        "description": "QA Checklist Rest 5",
        "id": 5,
      },
      Object {
        "description": "QA Checklist Rest 6",
        "id": 6,
      },
    ],
  },
  Object {
    "id": 1,
    "name": "Onboarding Rest",
    "tasks": Array [
      Object {
        "description": "Onboarding REST 1",
        "id": 1,
        "user": Object {
          "id": 8,
          "name": "Joe",
        },
      },
      Object {
        "description": "Onboarding REST 2",
        "id": 2,
      },
    ],
  },
  Object {
    "id": 2,
    "name": "Project Audit Rest",
    "tasks": Array [
      Object {
        "description": "Project Audit Rest 3",
        "id": 3,
      },
      Object {
        "description": "Project Audit Rest 4",
        "id": 4,
      },
    ],
  },
]
`;

exports[`Nested resources includes with hasMany.hasMany 1`] = `
Array [
  Object {
    "id": 11,
    "specData": "Should show up",
    "specDetails": Array [
      Object {
        "coms": Array [
          Object {
            "comData": "Should show up",
            "id": 2,
          },
        ],
        "id": 12,
        "specDetailData": "Should show up 1",
      },
      Object {
        "id": 13,
        "specDetailData": "Should show up 2, no coms",
      },
    ],
  },
]
`;
